src/multiversioning.zig

```diff
@@ -0,0 +1,836 @@
+        }
+
+        return self;
+    }
+
+    pub fn calculate_metadata_checksum(self: *const MultiversionMetadata) u128 {
+        // Checksum must have been set by this point.
+        assert(self.checksum_without_metadata != 0);
+
+        const self_without_checksum = std.mem.asBytes(
+            self,
+        )[0 .. @sizeOf(MultiversionMetadata) - @sizeOf(u128)];
+
+        return checksum.checksum(self_without_checksum);
+    }
+
+    /// Given a version, return all the versions:
+    /// * Older than the specified from_version,
+    /// * Newer than the current from_version, up to and including a newer one with the `visits`
+    ///   flag set.
+    // FIXME: Add test
+    pub fn advertisable(self: *const MultiversionMetadata, from_version: Release) ReleaseList {
+        var release_list: ReleaseList = .{};
+
+        for (0..self.past.count) |i| {
+            release_list.append_assume_capacity(Release{ .value = self.past.versions[i] });
+            if (self.past.versions[i] <= from_version.value) {
+                continue;
+            } else if (self.past.visits[i]) {
+                break;
+            }
+        }
+
+        // FIXME: Add current_version ONLY if needed
+        release_list.append_assume_capacity(Release{ .value = self.current_version });
+
+        assert(release_list.count() > 0);
+        return release_list;
+    }
+
+    comptime {
+        // Changing these will affect the structure stored on disk, which has implications for past
+        // clients trying to read!
+        assert(constants.vsr_releases_max == 64);
+        assert(PastVersionPack.past_releases_max == 63);
+        assert(@sizeOf(MultiversionMetadata) == 18288);
+        assert(@offsetOf(MultiversionMetadata, "checksum_metadata") == 18272);
+    }
+};
+
+const memfd_path = "tigerbeetle-multiversion";
+
+pub const Multiversion = struct {
+    pub const Callback = *const fn (*Multiversion, anyerror!void) void;
+    const ArgsEnvp = struct {
+        // Coerces to [*:null]const ?[*:0]const u8 but lets us keep information to free the memory
+        // later.
+        args: [:null]?[*:0]const u8,
+        envp: [*:null]const ?[*:0]const u8,
+    };
+
+    io: *IO,
+
+    exe_path: [:0]const u8,
+    args_envp: ArgsEnvp,
+
+    source_buffer: []u8,
+    source_fd: ?os.fd_t = null,
+
+    target_fd: os.fd_t,
+    target_pack_offset: ?usize = null,
+    target_metadata: ?MultiversionMetadata = null,
+    releases_bundled: ReleaseList = .{},
+
+    completion: IO.Completion = undefined,
+
+    timeout_completion: IO.Completion = undefined,
+    timeout_statx: os.linux.Statx = undefined,
+    timeout_statx_previous: union(enum) { none, previous: os.linux.Statx, err } = .none,
+    timeout_enabled: bool = true,
+
+    stage: union(enum) {
+        init,
+
+        source_stat,
+        source_open,
+        source_read,
+        source_swap,
+
+        ready,
+        err: anyerror,
+    } = .init,
+
+    callback: ?Callback = null,
```

Comment:
`callback` doesn't seem to be used for much... just

```
    fn on_read_from_binary_statx(self: *Multiversion, result: anyerror!void) void {
        self.start_timeout();

        _ = result catch return;
    }
```

Imo it would be simpler to remove `callback` and just hardcode `self.start_timeout()`.

Also, I think that would mean you could remove the `self.start_timeout` from `fn start()`, though maybe that messes up the timing.

Reply:
Good idea. With `timeout_start_enabled` I've gone the route of disabling timeouts at the beginning of `start()`, and explicitly starting them when its finished to ensure the timing is correct.